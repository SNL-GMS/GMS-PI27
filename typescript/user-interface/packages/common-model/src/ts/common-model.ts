import * as ArrayUtil from './array-util';
import * as BeamformingTemplateTypes from './beamforming-templates/types';
import * as ChannelSegmentTypes from './channel-segment';
import * as ColorTypes from './color/types';
import * as CommonTypes from './common';
import * as Displays from './displays/types';
import * as Endpoints from './endpoints/types';
import * as EventTypes from './event';
// required for legacy components
import * as LegacyEventTypes from './event/legacy';
import * as FacetedTypes from './faceted';
import * as FilterUtil from './filter/filter-util';
import * as FilterTypes from './filter/types';
import * as FkTypes from './fk';
import * as ProcessingMaskDefinitionTypes from './processing-mask-definitions/types';
import * as QcSegmentTypes from './qc-segment';
import * as RotationTypes from './rotation/types';
import * as SignalDetectionTypes from './signal-detection';
import * as ChannelTypes from './station-definitions/channel-definitions';
import * as ResponseTypes from './station-definitions/response-definitions';
import * as StationTypes from './station-definitions/station-definitions/station-definitions';
import * as ProcessingStationTypes from './station-processing/types';
import * as ReferenceStationTypes from './station-reference/types';
import * as TimeTypes from './time';
import * as TypeUtil from './type-util/type-util';
import * as ConfigurationTypes from './ui-configuration/types';
import * as UserProfileTypes from './user-profile/types';
import * as WaveformTypes from './waveform/types';
import * as WaveformUtil from './waveform/util';
import * as WorkflowTypes from './workflow/types';

export * from './faceted/utils';

export {
  ArrayUtil,
  BeamformingTemplateTypes,
  ChannelSegmentTypes,
  ChannelTypes,
  ColorTypes,
  CommonTypes,
  ConfigurationTypes,
  Displays,
  Endpoints,
  EventTypes,
  FacetedTypes,
  FilterTypes,
  FilterUtil,
  FkTypes,
  LegacyEventTypes,
  ProcessingMaskDefinitionTypes,
  ProcessingStationTypes,
  QcSegmentTypes,
  ReferenceStationTypes,
  ResponseTypes,
  RotationTypes,
  SignalDetectionTypes,
  StationTypes,
  TimeTypes,
  TypeUtil,
  UserProfileTypes,
  WaveformTypes,
  WaveformUtil,
  WorkflowTypes
};
