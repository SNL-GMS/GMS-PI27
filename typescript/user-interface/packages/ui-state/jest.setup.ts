/* eslint-disable max-classes-per-file */
/* eslint-disable import/first */
/* eslint-disable import/no-extraneous-dependencies */
import 'fake-indexeddb/auto';

import type { FilterDefinition } from '@gms/common-model/lib/filter/types';
import crypto from 'crypto';
import * as util from 'util';

import {
  channelSegmentWithSamples,
  valuesAsNumbers
} from './__tests__/__data__/ui-channel-segments/ui-channel-segment-data';

// This is to fix an issue with test leaks related to createStateSyncMiddleware() in jsdom
process.env.GMS_DISABLE_REDUX_STATE_SYNC = 'true';

process.env.GMS_DISABLE_PRE_CACHE_CHANNEL_SEGMENT_BY_CHANNEL = 'false';

Object.defineProperty(global, 'crypto', {
  value: {
    subtle: {
      digest: async (algorithm: AlgorithmIdentifier, msg: Uint8Array) =>
        (crypto.webcrypto as any).subtle.digest(algorithm, msg) as Promise<any>
    }
  }
});

Object.defineProperty(window, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(window, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});
Object.defineProperty(global, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(global, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});

import type { CommonTypes, WaveformTypes } from '@gms/common-model';
import { TimeseriesType } from '@gms/common-model/lib/channel-segment';
import type { GlobalWithFetchMock } from 'jest-fetch-mock';

const customGlobal: GlobalWithFetchMock = global as unknown as GlobalWithFetchMock;
// eslint-disable-next-line @typescript-eslint/no-require-imports
customGlobal.fetch = require('jest-fetch-mock');

customGlobal.fetchMock = customGlobal.fetch;

// eslint-disable-next-line @typescript-eslint/no-require-imports
require('jest-canvas-mock');

const globalAny: any = global;

// eslint-disable-next-line @typescript-eslint/no-require-imports
globalAny.fetch = require('jest-fetch-mock');

globalAny.TextEncoder = util.TextEncoder;

globalAny.window = window;
globalAny.console = {
  ...globalAny.console,
  warn: jest.fn()
};

class Worker {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public constructor(stringUrl: string) {
    this.url = stringUrl;
    this.onmessage = () => {
      // do nothing
    };
  }

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}
globalAny.window.Worker = Worker;

class MessagePort {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}

class SharedWorker {
  public url: string;

  public port = new MessagePort();

  public constructor(stringUrl: string) {
    this.url = stringUrl;

    this.port.onmessage = () => {
      // do nothing
    };
  }
}
globalAny.window.SharedWorker = SharedWorker;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
require('jest-fetch-mock').enableMocks();

jest.mock('@gms/ui-wasm', () => {
  const isDesigned = jest.requireActual('@gms/ui-wasm/lib/gms-interop/is-designed');

  const filter = async (
    filterDefinition: FilterDefinition,
    data: Float64Array
  ): Promise<Float64Array> => Promise.resolve(data);

  const design = async (filterDefinition: FilterDefinition): Promise<FilterDefinition> =>
    Promise.resolve(filterDefinition);

  const getBoundsForPositionBuffer = async (): Promise<{
    min: number;
    minSecs: number;
    max: number;
    maxSecs: number;
  }> => Promise.resolve({ min: 0, minSecs: 0, max: 10, maxSecs: 10 });

  const waveformNorth: WaveformTypes.Waveform = {
    samples: new Float64Array(valuesAsNumbers),
    type: TimeseriesType.WAVEFORM,
    startTime: 0,
    endTime: 10,
    sampleRateHz: 40,
    sampleCount: valuesAsNumbers.length
  };

  const waveformEast: WaveformTypes.Waveform = {
    samples: new Float64Array(valuesAsNumbers),
    type: TimeseriesType.WAVEFORM,
    startTime: 11,
    endTime: 20,
    sampleRateHz: 40,
    sampleCount: valuesAsNumbers.length
  };

  const maskAndRotate2d = async (): Promise<
    CommonTypes.TimeseriesWithMissingInputChannels<WaveformTypes.Waveform>[]
  > =>
    Promise.resolve([
      {
        timeseries: [waveformNorth],
        missingInputChannels: []
      },
      {
        timeseries: [waveformEast],
        missingInputChannels: []
      }
    ]);

  return {
    design,
    filter,
    maskAndRotate2d,
    getBoundsForPositionBuffer,
    isDesigned: isDesigned.isDesigned,
    // eslint-disable-next-line @typescript-eslint/require-await
    maskAndBeamWaveforms: jest.fn(async () => channelSegmentWithSamples)
  };
});

// eslint-disable-next-line @typescript-eslint/no-magic-numbers
Date.now = jest.fn().mockReturnValue(250);
