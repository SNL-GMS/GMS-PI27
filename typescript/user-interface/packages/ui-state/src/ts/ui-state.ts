export * from './app';
export * from './imperative-context-menu';
export * from './types';
export * from './workers';
