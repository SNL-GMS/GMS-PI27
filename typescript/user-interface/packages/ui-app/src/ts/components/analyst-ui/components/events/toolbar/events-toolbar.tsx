import { Displays } from '@gms/common-model';
import { ButtonToolbarItem, CheckboxDropdownToolbarItem, Toolbar } from '@gms/ui-core-components';
import {
  DisplayedEventsConfigurationEnum,
  eventsActions,
  EventsColumn,
  selectDisplayedEventsConfiguration,
  selectEventsColumnsToDisplay,
  useAppDispatch,
  useAppSelector
} from '@gms/ui-state';
import Immutable from 'immutable';
import React from 'react';

import type { CreateEventMenuState } from '~analyst-ui/common/menus/create-event-menu-item';
import { useCreateEventControl } from '~analyst-ui/common/toolbar-items/create-event-control';
import { systemConfig } from '~analyst-ui/config/system-config';
import { useSelectionInformationControl } from '~common-ui/common/selection-information';
import { convertMapToObject, convertObjectToEventsColumnMap } from '~common-ui/common/table-utils';
import { useBaseDisplaySize } from '~common-ui/components/base-display/base-display-hooks';

import { setFocusToEventsDisplay } from '../events-util';
import {
  displayedEventsDropdownDisplayStrings,
  displayedEventsLabelStrings,
  displayedEventsRenderDividers,
  eventColumnDisplayStrings
} from '../types';
import type { CountFilterOptions, EventCountEntry } from './event-count-toolbar-item';
import { useEventsCountToolbarItem } from './event-count-toolbar-item';

export interface EventsToolbarProps {
  readonly countEntryRecord: Record<CountFilterOptions, EventCountEntry>;
  readonly disableMarkSelectedComplete: boolean;
  handleMarkSelectedComplete(): void;
  setCreateEventMenuState: (value: CreateEventMenuState) => void;
}

export function EventsToolbar({
  countEntryRecord,
  disableMarkSelectedComplete,
  handleMarkSelectedComplete,
  setCreateEventMenuState
}: EventsToolbarProps) {
  const [widthPx] = useBaseDisplaySize();

  const dispatch = useAppDispatch();
  const eventsColumnsToDisplayObject = useAppSelector(selectEventsColumnsToDisplay);
  const columnsToDisplay = React.useMemo(
    () => convertObjectToEventsColumnMap(eventsColumnsToDisplayObject),
    [eventsColumnsToDisplayObject]
  );

  const displayedEventsConfiguration = useAppSelector(selectDisplayedEventsConfiguration);

  const createEventControl = useCreateEventControl(setCreateEventMenuState, 'evntcreateevent');
  const eventsCountToolbarItem = useEventsCountToolbarItem('events count', countEntryRecord);
  const selectionInformation = useSelectionInformationControl(
    'eventselectioninfo',
    Displays.IanDisplays.EVENTS,
    setFocusToEventsDisplay
  );

  const toolbarItemsLeft: JSX.Element[] = React.useMemo(
    () => [eventsCountToolbarItem, selectionInformation],
    [eventsCountToolbarItem, selectionInformation]
  );

  // Map of values to pass to the checkbox dropdown toolbar item
  const displayedEventsConfigurationMap: Immutable.Map<DisplayedEventsConfigurationEnum, boolean> =
    React.useMemo(
      () => Immutable.fromJS(displayedEventsConfiguration),
      [displayedEventsConfiguration]
    );

  const toolbarItemsRight: JSX.Element[] = React.useMemo(() => {
    const setColumnsToDisplay = (cols: Immutable.Map<EventsColumn, boolean>) =>
      dispatch(eventsActions.updateEventsColumns(convertMapToObject(cols)));

    const setEventsToDisplay = (
      events: Immutable.Map<DisplayedEventsConfigurationEnum, boolean>
    ) => {
      dispatch(eventsActions.updateDisplayedEventsConfiguration(convertMapToObject(events)));
    };
    return [
      <CheckboxDropdownToolbarItem
        key="shownevents"
        checkboxEnum={DisplayedEventsConfigurationEnum}
        label="Show Events"
        widthPx={150}
        tooltip="Set which edge events are visible"
        enumToCheckedMap={displayedEventsConfigurationMap}
        enumKeysToDisplayStrings={displayedEventsDropdownDisplayStrings}
        enumKeysToDividerMap={displayedEventsRenderDividers}
        enumKeysToLabelMap={displayedEventsLabelStrings}
        onChange={setEventsToDisplay}
        cyData="filter-column"
      />,
      <CheckboxDropdownToolbarItem
        key="showncolumns"
        checkboxEnum={EventsColumn}
        enumKeysToDisplayStrings={eventColumnDisplayStrings}
        label="Show Columns"
        widthPx={150}
        tooltip="Set which columns are visible"
        enumToCheckedMap={columnsToDisplay}
        onChange={setColumnsToDisplay}
        cyData="filter-column"
      />,
      createEventControl,
      <ButtonToolbarItem
        key="markcomplete"
        cyData="mark-open-complete"
        tooltip="Mark selected events complete"
        label="Mark Selected Complete"
        onButtonClick={handleMarkSelectedComplete}
        disabled={disableMarkSelectedComplete}
      />
    ];
  }, [
    columnsToDisplay,
    disableMarkSelectedComplete,
    displayedEventsConfigurationMap,
    createEventControl,
    handleMarkSelectedComplete,
    dispatch
  ]);

  return (
    <Toolbar
      toolbarWidthPx={widthPx}
      parentContainerPaddingPx={systemConfig.marginForToolbarPx}
      itemsLeft={toolbarItemsLeft}
      itemsRight={toolbarItemsRight}
    />
  );
}
