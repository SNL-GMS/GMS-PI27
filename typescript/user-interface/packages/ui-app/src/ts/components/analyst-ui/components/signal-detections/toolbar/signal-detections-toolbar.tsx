import { Displays } from '@gms/common-model';
import { CheckboxDropdownToolbarItem, Toolbar } from '@gms/ui-core-components';
import type { AppDispatch } from '@gms/ui-state';
import {
  DisplayedSignalDetectionConfigurationEnum,
  selectDisplaySignalDetectionConfiguration,
  SignalDetectionColumn,
  signalDetectionsActions,
  useAppDispatch,
  useAppSelector
} from '@gms/ui-state';
import Immutable from 'immutable';
import React from 'react';

import { useCreateEventControl } from '~analyst-ui/common/toolbar-items/create-event-control';
import type { SignalDetectionsToolbarProps } from '~analyst-ui/components/signal-detections/types';
import {
  signalDetectionColumnDisplayStrings,
  signalDetectionSyncDisplayStrings,
  signalDetectionSyncLabelStrings,
  signalDetectionSyncRenderDividers
} from '~analyst-ui/components/signal-detections/types';
import { useSelectionInformationControl } from '~common-ui/common/selection-information';
import { useBaseDisplaySize } from '~common-ui/components/base-display/base-display-hooks';

import { systemConfig } from '../../../config/system-config';
import { setFocusToSignalDetectionDisplay } from '../table/signal-detections-table-utils';
import { useSignalDetectionsCountToolbarItem } from './signal-detection-count-toolbar-item';

/**
 * Referentially-stable onChange handler function for the "Show Detections"
 * checkbox item
 */
export const showDetectionsOnChange =
  (reduxDispatch: AppDispatch) =>
  (event: any): void => {
    const eventObjectString = JSON.stringify(event);
    reduxDispatch(
      signalDetectionsActions.updateDisplayedSignalDetectionConfiguration(
        JSON.parse(eventObjectString)
      )
    );
  };

/**
 * Toolbar for the Signal Detections list display. Contains intractable elements
 * like buttons, checkboxes that can change the behavior/appearance of the display.
 */
export function SignalDetectionsToolbar(props: SignalDetectionsToolbarProps) {
  const {
    countEntryRecord,
    selectedSDColumnsToDisplay,
    setSelectedSDColumnsToDisplay,
    setCreateEventMenuState
  } = props;

  const [displayWidthPx] = useBaseDisplaySize();
  const dispatch = useAppDispatch();
  const displayedSignalDetectionConfigurationObject = useAppSelector(
    selectDisplaySignalDetectionConfiguration
  );

  const signalDetectionsCountToolbarItem = useSignalDetectionsCountToolbarItem(
    'signal detections count',
    countEntryRecord
  );
  const createEventControl = useCreateEventControl(setCreateEventMenuState, 'sdcreateevent');
  const selectionInformation = useSelectionInformationControl(
    'sdselectioninfo',
    Displays.IanDisplays.SIGNAL_DETECTIONS,
    setFocusToSignalDetectionDisplay
  );

  // Map of values to pass to the checkbox dropdown toolbar item
  const displayedSignalDetectionConfigurationMap: Immutable.Map<
    DisplayedSignalDetectionConfigurationEnum,
    boolean
  > = React.useMemo(
    () => Immutable.fromJS(displayedSignalDetectionConfigurationObject),
    [displayedSignalDetectionConfigurationObject]
  );

  const itemsLeft: JSX.Element[] = React.useMemo(
    () => [signalDetectionsCountToolbarItem, selectionInformation],
    [signalDetectionsCountToolbarItem, selectionInformation]
  );

  const itemRight: JSX.Element[] = React.useMemo(
    () => [
      <CheckboxDropdownToolbarItem
        key="SDdetections"
        label="Show Detections"
        menuLabel="Show Detections"
        tooltip="Set which detections are visible"
        cyData="sd-table-detections-toggle"
        enumToCheckedMap={displayedSignalDetectionConfigurationMap}
        checkboxEnum={DisplayedSignalDetectionConfigurationEnum}
        enumKeysToDisplayStrings={signalDetectionSyncDisplayStrings}
        enumKeysToDividerMap={signalDetectionSyncRenderDividers}
        enumKeysToLabelMap={signalDetectionSyncLabelStrings}
        onChange={showDetectionsOnChange(dispatch)}
      />,
      <CheckboxDropdownToolbarItem
        key="SDcolumns"
        label="Show Columns"
        menuLabel="Show Columns"
        tooltip="Set which columns are visible"
        onChange={setSelectedSDColumnsToDisplay}
        cyData="sd-table-column-picker"
        enumToCheckedMap={selectedSDColumnsToDisplay}
        checkboxEnum={SignalDetectionColumn}
        enumKeysToDisplayStrings={signalDetectionColumnDisplayStrings}
        itemSide="LEFT"
      />,
      createEventControl
    ],
    [
      createEventControl,
      dispatch,
      displayedSignalDetectionConfigurationMap,
      selectedSDColumnsToDisplay,
      setSelectedSDColumnsToDisplay
    ]
  );

  return (
    <Toolbar
      toolbarWidthPx={displayWidthPx}
      parentContainerPaddingPx={systemConfig.marginForToolbarPx}
      itemsLeft={itemsLeft}
      itemsRight={itemRight}
    />
  );
}
