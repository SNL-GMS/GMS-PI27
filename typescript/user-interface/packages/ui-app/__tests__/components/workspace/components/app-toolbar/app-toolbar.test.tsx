/* eslint-disable react/jsx-props-no-spreading */
import { UserProfileTypes } from '@gms/common-model';
import { getStore } from '@gms/ui-state';
import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import * as DeprecatedToolbar from '../../../../../src/ts/components/workspace/components/toolbar/toolbar-component';
import type { DeprecatedToolbarProps } from '../../../../../src/ts/components/workspace/components/toolbar/types';

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

describe('App Toolbar', () => {
  const testProps: DeprecatedToolbarProps = {
    components: null,
    saveDialog: null,
    setLayout: null,
    logo: 'test_logo',
    userName: 'test_name',
    isAboutDialogOpen: false,
    isSaveWorkspaceAsDialogOpen: false,
    openLayoutName: 'default_layout_name',
    versionInfo: {
      versionNumber: 'string',
      commitSHA: 'string'
    },
    userProfile: {
      userId: 'userId',
      defaultAnalystLayoutName: 'default_layout_name',
      preferences: {
        currentTheme: 'GMS Dark Mode',
        colorMap: 'turbo'
      },
      workspaceLayouts: [
        {
          name: 'default_layout_name',
          layoutConfiguration: 'asdf',
          supportedUserInterfaceModes: [UserProfileTypes.UserMode.IAN]
        }
      ]
    },
    setOpenLayoutName: (): any => {
      jest.fn();
    },
    clearLayout: () => jest.fn(),
    getOpenDisplays: () => [],
    logout: () => jest.fn(),
    openDisplay: () => {
      jest.fn();
    },
    openWorkspace: () => {
      jest.fn();
    },
    showLogPopup: () => jest.fn(),
    showAboutDialog: () => jest.fn(),
    toggleSaveWorkspaceAsDialog: () => jest.fn()
  };

  const { container } = render(
    <Provider store={getStore()}>
      <DeprecatedToolbar.DeprecatedToolbar {...testProps} />
    </Provider>
  );

  it('renders a component', () => {
    expect(container).toMatchSnapshot();
  });

  it('can reset an animation', () => {
    const target = {
      current: {
        className: '',
        offsetWidth: -1
      }
    };
    DeprecatedToolbar.triggerAnimation(target as React.MutableRefObject<HTMLSpanElement>);
    expect(target.current.className).toEqual('workspace-logo__label keypress-signifier');
  });
});
