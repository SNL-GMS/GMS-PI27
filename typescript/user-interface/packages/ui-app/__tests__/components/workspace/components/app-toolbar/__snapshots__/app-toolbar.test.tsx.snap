// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Toolbar renders a component 1`] = `
<div>
  <nav
    class="bp5-navbar workspace-navbar"
  >
    <div
      class="workspace-navbar__image-label"
    >
      <img
        alt="GMS Logo"
        class="workspace-logo"
        data-cy="workspace-logo"
        height="33"
        src="test_logo"
        width="35"
      />
      <span
        class="workspace-logo__label keypress-signifier"
      >
        GMS
      </span>
    </div>
    <span
      aria-expanded="false"
      aria-haspopup="true"
      class="bp5-popover-target"
    >
      <button
        class="bp5-button bp5-minimal app-menu-button"
        data-cy="username"
        type="button"
      >
        <span
          class="bp5-button-text"
        >
          test_name
        </span>
        <span
          aria-hidden="true"
          class="bp5-icon bp5-icon-standard bp5-icon-user"
          data-icon="user"
        />
      </button>
    </span>
  </nav>
</div>
`;
